src/spartan_handler_fsm.erl

```diff
@@ -156,15 +135,9 @@ execute(timeout, State = #state{data = Data}) ->
             spartan_metrics:update([?APP, no_upstreams_available], 1, ?SPIRAL),
             reply_fail(State1),
             {stop, normal, State};
-        Upstreams0 ->
+        Upstreams ->
             StartTimestamp = os:timestamp(),
-            QueryUpstreams = take_upstreams(Upstreams0),
-            OutstandingUpstreams = lists:map(fun(Upstream) ->
-                            Pid = spawn_link(?MODULE,
-                                             resolve,
-                                             [self(), Upstream, State]),
-                            {Upstream, Pid}
-                            end, QueryUpstreams),
+            OutstandingUpstreams = start_workers(Upstreams, State),
```

Comment:
Isn't process creation cheap? If there is not much gain or we are unsure then we should keep the old design.